### ISSUE TITLE
Improve search relevance

### ISSUE THREAD

**Issue by heathermiller** [scalacenter/scaladex#389]
I can't find spark anymore... I have to search for Spark and then fiddle with the filters. Spark should really turn up on the first page when searching for "spark"

**Comment by julienrf** [comment]
The score is highly dependent on keywords, I think that's the best way to avoid false positives, even though it requires manual editing of projects. 

The other results for the "spark" query are not irrelevant. It is super hard to design a magic scoring function that always put your expected result first... 

I added the "spark" keyword to the apache/spark project, in order to boost its score.

**Comment by kzys** [comment]
Can we just use the number of stars as the default sort order?
https://index.scala-lang.org/search?q=spark&page=1&sort=stars

**Comment by julienrf** [comment]
@kzys We tried that but then some projects that are not completely related to a given search query but have a lot of stars (e.g. playframework) appear at the top of the results. That does not work well either…

**Comment by MasseGuillaume** [comment]
@heathermiller let's use this issue: https://github.com/scalacenter/scaladex/issues/303 and close this one as duplicate ?

**Comment by cvogt** [comment]
this looks off
![image](https://cloud.githubusercontent.com/assets/274947/23582608/1df76388-00fc-11e7-95eb-4ee28f295f13.png)

**Comment by cvogt** [comment]
scalafmt doesn't show anywhere in this list eventhough I typed the exact name

**Comment by cvogt** [comment]
when I type enter it does show on position 8 of that list.

maybe project that have the exact search term contained in there name should show first?

**Comment by olafurpg** [comment]
Scalafmt does show up when searching for scalameta

<img width="860" alt="screen shot 2017-03-29 at 14 18 07" src="https://cloud.githubusercontent.com/assets/1408093/24454070/a31c7aac-148a-11e7-9a3a-ec29d55c4175.png">

I think the scalameta repo should be the top result, however.

**Comment by ShaneDelmore** [comment]
Can we give a heavy weight to matches in the name?  If I search for shapeless I should get shapeless, even if a library named ShapetyShapeShape mentions Shapeless 42 times in the read even and has more stars than Travis Brown has StackOverflow points. Is this controversial?

**Comment by MasseGuillaume** [comment]
@ShaneDelmore The counter example is a library named json. Is it more relevant than circe ?

**Comment by ShaneDelmore** [comment]
Yes. If I search for a library named json I would like it to show up before Circe. If I list the category json by popularity then I would expect Circe to be near the top. Has the counter example proved a larger problem than the current results ranking?

**Comment by olafurpg** [comment]
The search results for "scalafix" lists scalacenter/scalafix at 11th place https://index.scala-lang.org/search?q=scalafix

<img width="780" alt="screen shot 2017-06-14 at 09 11 10" src="https://user-images.githubusercontent.com/1408093/27119787-7f0250f4-50e1-11e7-98ef-fe9a92582eb1.png">

**Comment by heathermiller** [comment]
This issue of search relevance really needs to be resolved. This is crippling Scaladex.

**Comment by olafurpg** [comment]
Searching for "shapeless" does not list https://github.com/milessabin/shapeless on the front page https://index.scala-lang.org/search?q=shapeless at the moment.

**Comment by MasseGuillaume** [comment]
And someone lost his cat: https://github.com/scalacenter/scaladex/issues/430.

I'm taking a look.

**Comment by MasseGuillaume** [comment]
@olafurpg found the problem:

**Comment by MasseGuillaume** [comment]
I added [test for relevance](https://github.com/scalacenter/scaladex/blob/master/server/src/test/scala/ch.epfl.scala.index.server/RelevanceTest.scala#L21)

[`first(query)(org/repo)`](https://github.com/scalacenter/scaladex/blob/398eb176d944c6476ffb5394d6f3a65252ec37b3/server/src/test/scala/ch.epfl.scala.index.server/RelevanceTest.scala#L37): org/repo should be the first result when searching for query
[`top(query)(repos)`](https://github.com/scalacenter/scaladex/blob/398eb176d944c6476ffb5394d6f3a65252ec37b3/server/src/test/scala/ch.epfl.scala.index.server/RelevanceTest.scala#L66): repos should be on the first page the ordering is not relevant
[`exactly(query)(repos)`](https://github.com/scalacenter/scaladex/blob/398eb176d944c6476ffb5394d6f3a65252ec37b3/server/src/test/scala/ch.epfl.scala.index.server/RelevanceTest.scala#L98) repos should show in this order

**Comment by ShaneDelmore** [comment]
I attempted to work on this a couple of weekends ago and then forgot about it, and now I am out of motivation for the time being but I’ll share my thoughts anyway for what I found before I stopped due to not knowing how to inspect the contents of the index to see what to search on.

Multiple people have said something to the effect of “just rank by github stars” and the answer has often been “you don’t want to do that, or spark will always be at the top of the list”.  I believe the problem is that there is not enough separation between filtering results returned and ranking them.

The ranking isn’t bad from what I can see although I did improve results in my tests by playing with the weights a bit, but the biggest difference was to just stop trying so hard to make sure we never miss a result.  I noticed that we don’t just search title, tags, and the readme, but we search dependency names, and other things that I think should not cause a library to be returned as a result.

For example, if you search for config and a library that does Units of Measure calculations ends up depending on typesafe-config, or maybe mentions that you can set defaults in the lib.config file it will be included as a result.  I don’t think anyone wants this (please post here if I am wrong though, maybe I don’t know of your use case).  

If we stop including so many libraries in results, then we won’t have to worry about them getting promoted to the top of the list.  The problem is not that when querying json that spark has more github stars than circe, the problem is that Spark is not a json library.  Just having a dependency on a module named spark-json doesn’t make it a json library.  Heck, if we want to match spark-json, great, return that as a hit result, but not spark itself.  

Out of ~6000 libraries I get the following number of hits for various searches:
Scalatest: 2026
Json: 1106
Play: 989
Config: 852
Shapeless: 323

I think these numbers are hugely exaggerated, almost all due to dependencies.  Due we have 323 libraries that depend on shapeless, sure, but we only have a few libs that are the one and only Shapeless or part of it’s ecosystem and while a lot of libraries use scalatest, when someone searches for scalatest they want ScalaTest and libraries that add functionality to ScalaTest, not every library that happens to use it in it’s own tests.

TLDR: If we return fewer, more targeted results I think the ordering will end up being pretty easy, I think we are just returning far too many results.  Don’t use dependency names for inclusion criteria.  The problem is not missing a few libs, if libs are missed the author can just add a topic with the search term they want to hit, but if we include too many libs then it becomes difficult to rank them and the lib author has far less control over that.

I’m sorry I’m just leaving a comment and not a PR but I’m realizing that I may not have time to get to this for a while and my findings may be useful for someone else.

**Comment by eed3si9n** [comment]
It's now 2019. And clicking on the first topic from the front page [akka (110)](https://index.scala-lang.org/search?q=*&topics=akka), we still get odd results.

<img width="1210" alt="topics_akka_2019" src="https://user-images.githubusercontent.com/184683/57388150-a38c9800-7185-11e9-969c-ee8e6b7a74f5.png">

[json(85)](https://index.scala-lang.org/search?q=*&topics=json) not much better since it missing obvious ones like circe and json4s on the first page.

<img width="1229" alt="topics_json_2019" src="https://user-images.githubusercontent.com/184683/57388445-2c0b3880-7186-11e9-9f99-979b2e39a734.png">

In both cases GitHub's relevance search filtered by Scala language returns much usable results:

<img width="1217" alt="2019_Search_akka" src="https://user-images.githubusercontent.com/184683/57388776-eac75880-7186-11e9-96d4-3130da4fd76e.png">
<img width="1229" alt="2019_Search_json" src="https://user-images.githubusercontent.com/184683/57388885-2104d800-7187-11e9-9e78-3e670d79338b.png">

From the looks of it it's a mixture of repo name, tags, description, and stars?

**Comment by eed3si9n** [comment]
Here's a back of the napkin weights I reverse engineered from looking at the ranking.
https://docs.google.com/spreadsheets/d/1L6IjOKJ67GjwiBdLGa8M7GSa_AU645WmtYpkXN4ujeY/edit#gid=0

<img width="994" alt="GitHub_relevance" src="https://user-images.githubusercontent.com/184683/57395672-b52a6b80-7196-11e9-9f71-9441ee5d1cae.png">

On the "name" column you get 10 points for exact match on the project name like "akka", otherwise you get 1 point. paypal/squbs showing up at the second place is a hint of how the repo names are weighed against the stars.

akka-sample showing up above akka-in-action shows that tags do matter, but proportionally, it should matter much _less_ compared to the description.

**Comment by eed3si9n** [comment]
I attempted to send a PR to implement the scoring formula - https://github.com/scalacenter/scaladex/pull/571
If someone could review and correct it that'd be great.

**Comment by eed3si9n** [comment]
I also agree with @ShaneDelmore's comment about

> I think these numbers are hugely exaggerated, almost all due to dependencies. 

Could someone show me how to cut those out?

**Comment by adpi2** [comment]
Hello everyone,

I have made a first attempt to improve the scaladex search enging in this PR #597, that also ships the changes that @eed3si9n proposed last year. I would like people to test the result of this PR. I have deployed it in the staging environment so that you can compare with the current behaviour.
- staging: [https://index-dev.scala-lang.org](https://index-dev.scala-lang.org)
- prod: [https://index.scala-lang.org](https://index.scala-lang.org)

Let me know what your overall feeling is about it. Does it seem better that the current search engine ? Have you found some inconsistencies ? Have you found some behavior regression ?

## Description

### Reducing the number of results
This is mostly the result of not searching in the dependencies. I have compared the number of results for those queries:
| query  | master | PR |
|---|---|---|
| `scalatest` | 2862  | 167 |
| `shapeless` | 439 | 206 |
| `scalaz` | 7028 | 218 |
| `akka` | 1114 | 741 |

### Balancing the term matching and popularity scoring
I try to give equal importance to term matching and popularity. A project needs both to appear in the first page. That's why I changed the scoring combining function from `add` to `multiply`. 
I used a logarithmic scale for smoothing the popularity scoring, so that the more popular projects don't appear first when the match scoring is not good enough.
When searching for `json` the 5 top results are now, in order:
- `spray/spray-json`
- `circe/circe`
- `nestorpersist/json`
- `json4s/json4s`
- `playframework/play-json`

### Boosting the github description and github readme
Some popular projects don't have explicit name and they don't use github tags. For instance, the term  `testing` only appears in the description of `scalatest`, the term `database` only appear in the readme of `slick`. I have chosen those boost values:
|field|boost|
|-|-|
|repository|16|
|primaryTopic|8|
|github.description|4|
|organization|4|
|github.topics|4|
|artifacts|2|
|github.readme|1|

### Using full text queries
When entering several terms, the user might be interseted in results that combine some of those terms, the more the better. Exact match queries return only a few results, mostly unpopular. Full text queries return much more result and sort them by relevance and popularity.

## Review
You may want to try those patterns:
- looking for a project: "scalatest", "akka", "spark", "doobie", "zio"
- searching a topic: "json", "web", "testing", "compiler", "distributed"
- searching a combination of terms: "kafka cassandra akka", "spark machine learning" 
- looking for a organization or a library author: "scalacenter", "typelevel", "lihaoyi"
- looking for an artifact inside a project: "scalctic"

**Comment by sjrd** [comment]
I like it! It feels much better now!

**Comment by julienrf** [comment]
Nice! The search results are great, but the search autocomplete is less usable than before :-/ For instance, you won’t get completion for shapeless unless you write “shapeless” (in production, you get completion as soon as you write “shapele”). That being said, this could be left for separate work.

**Comment by eed3si9n** [comment]
@adpi2 Thanks for moving this forward.

**Comment by adpi2** [comment]
I totally agree with you @julienrf. So far, the exact same query is used for autocomplete and actual search which is, in my opinion, not a good idea. I think most of the false positive in the actual search was due to the autocomplete feature and I removed them. That's why the autocomplete is not working so well anymore. I will try to separate those two features while keeping consistency.

**Comment by OlegYch** [comment]
I like it. Autocomplete behaviour doesn't seem like a problem. There is an existing problem with it that makes it hard to use though https://github.com/scalacenter/scaladex/issues/598

**Comment by adpi2** [comment]
A second batch of improvements have been deployed to [https://index.scala-lang.org/](https://index.scala-lang.org/)

Once again I would be more than happy if you could report any bug, issue or regression related to those changes.

## Description

### English language analysis

I have used the powerfull `english` elasticsearch analyzer which:
- is case insensitive
- removes all punctuation:
  - "unit-testing" matches with "unit testing"
  - "scala.js" and "scala-js" match "scala js"
- use a stemming algorithm:
  - "testing" and "tests" match with "test"
  - "parser" and "parsing" match with "parse"
- removes all frequent words: "it", "the", "a", "is"...

### `bestField` mode

I have used the `bestField` mode which only keeps the score of the field that matches best. The reason for this choice is that we don't want to penalize a project that has a good description but no tags compared to a project that has a good description, a primary tag and lot of tags.

## Results

| query | top results in order |
|-|-|
| tests | `typelevel/scalacheck` <br> `lihaoyi/utest` <br> `scalatest/scalatest` <br> `holdenk/spark-testing-base` <br> `gatling/gatling` <br> ... |
| database | `netflix/atlas` <br> `jooq/jooq` <br> `getquill/quill` <br> `neo4j/neo4j` <br> `tpolecat/doobie` <br> ... |
| json | `circe/circe` <br> `spray/spray-json` <br> `playframework/play-json` <br> `ngs-doo/dsl-json` <br> `json4s/json4s` <br> ... |
| scala javascript | `scala-js/scala-js` <br> ... |
| scala 3 | `lampepfl/dotty` <br> ... |

**Comment by eed3si9n** [comment]
For "json" the regular search is now better than topic feature. Maybe it should stop suggesting "Did you mean topics=json?", and also the front page topics should be changed to search terms?

**Comment by tpolecat** [comment]
Hi, I messed around with this for a bit and I agree the results are better but I think they're still too random to be useful for users who don't know what they're looking for. I don't think this is anyone's fault, it's just a very hard problem. I kind of wonder if we need something like "Editor's Picks" which highlight libraries that get a lot of mentions in the user survey or something.

Anyway here's an example. A search for `database` yields the following first page. My reaction to each item is obviously subjective but I have been working on database libraries in Scala for 7 years so I have some background in this corner of the ecosystem, although I have a blind spot for Spark/data things so some things that look weird to me might make sense.

1. `netflix/atlas` - never heard of it
1. `jooq/jooq` - this is a Java library with some Scala support, never heard anyone mention it in Scala channels
1. `getquill/quill` - :+1: good, should be on first page
1. `neo4j/neo4j` - this is a Java library with no Scala API
1. `tpolecat/doobie` - :+1: good, should be on first page
1. `lucidsoftware/relate` - relevant but not widely used as far as I know
1. `playframework/anorm` - 👍 relevant, widely used despite small number of stars, don't know how we could determine this from available metadata
1. `reactivemongo/reactivemongo` - 👍  good, should be on first page
1. `tpolecat/skunk` - this is experimental and would be filtered out if we discarded libraries with no 0.1+ release
1. `fakod/neo4j-scala` - relevant, would be nice if we could associate with `neo4j/neo4j` because there are a lot of Scala projects that provide support for underlying Java libs.
1. `scalikejdbc/scalikejdbc` - 👍  good, should be on first page
1. `simerplaha/swaydb` - relevant
1. `squeryl/squeryl` - relevant
1. `haifengl/unicorn` - never heard of it
1. `opencypher/openCypher` -  never heard of it
1. `rdbc-io/rdbc` - relevant
1. `outworkers/morpheus` - relevant, never heard of it
1. `kyleu/databaseflow` - relevant, never heard of it
1. `input-output-hk/iodb` - possibly relevant, never heard of it
1. `scalamolecule/molecule` - relevant, never heard of it

When I looked yesterday the top result was `rdbc-io / rdbc` so the results are shifting around I think (I assume you're still tweaking things).

So I think this is mostly ok although a new user may have a hard time figuring out which library to use. A big problem though is that 👉  **`slick/slick` doesn't appear until page 4** 👈 and it's probably the most widely-used database library in Scala right now.

Anyway thanks for the work on this. It's getting better but I think it's really a very hard problem that will continue to require fine-tuning.

**Comment by eed3si9n** [comment]
@tpolecat I was comparing the new results against GitHub's search results earlier, and I think "database" is particularly a tricky one because it could either be literal database like H2 or database accessing library like doobie. Here's GitHub - https://github.com/search?l=Scala&q=database&type=Repositories

Re: slick/slick, I think it needs to improve the description from "Scala Language Integrated Connection Kit" so something else for a search engine to pick it up.

> I kind of wonder if we need something like "Editor's Picks" which highlight libraries that get a lot of mentions in the user survey or something.

I do agree that supplementing search result with human curation is a good idea. It could even display a short paragraph of text recommending a few well-supported libraries for some obvious genres like test frameworks and web frameworks.

**Comment by adpi2** [comment]
Thank you @tpolecat for your valuable feedback and @eed3si9n for your contribution. There is still room for improvement and it is worth putting time and effort on this.

I think there are two challenges here. The first one is to come up to a good, reliable and easy to understand scoring algorithm for scala libraries. The second challenge is to encourage the library authors and maintainers to improve the SEO of their libraries. It would be super-easy, for instance, to boost the score of `slick/slick` by adding the `database` keyword in the github description or github tags of the repository. When it will be done, I am sure that `slick/slick` will reach the top of the list. A detailed `readme.md` also contributes to boosting a project.

Yet, you have revealed an interesting pattern in the `database` example, which is that native java projects are generally overrated by the socring function. Let's take the example of `neo4j/neo4j`:

- Why does `neo4j/neo4j` appear in scaladex? Because it contains a scala artifact called `neo4j-cypher-compiler-2.2`.
- Why does it appear in the search result for `database`? Because the `neo4j/neo4j` description mentions the `database` keyword`?
- Why does it appear on top of the list? Because it is a very popular project.
- Yet is it relevant? Not much, because `neo4j/neo4j` is a java project. A ridiculous fraction of its code is actually written in scala.

![neo4j-scala](https://user-images.githubusercontent.com/13123162/77322598-41dd0a00-6d14-11ea-84b4-a4e7df5fadff.jpeg)

We may be able to temper the score of such projects by combining it with the fraction of sala code. I am not sure though if this metric is available in the github apis.

Also, we may say that a project compiled with scala 2.13 is more valuable than a project compiled with scala 2.11. Of course there are exceptions such as spark, but I think this is generally true. What do you think ?

**Comment by tpolecat** [comment]
> It would be super-easy, for instance, to boost the score of slick/slick by adding the database keyword

Indeed, it looks like keywords end up being extremely valuable, score-wise (although I don't actually know which other factors you look at).

| Repo | Stars | Mentions of "database" in README | Keyword? | Page |
|----|---|--|--|--|
| `tpolecat/doobie` | 1.6k | 1 | yes | 1 |
| `slick/slick` | 2.3k | 7 | no |4 |

> Also, we may say that a project compiled with scala 2.13 is more valuable than a project compiled with scala 2.11.

Yes, I agree with this but I'm not a typical user. The user survey results might be a good place to get an accurate sense for this.

In general something that might be useful (but potentially overwhelming) would be to ask library authors to see if their own projects (and ones they're familiar with) are showing up with reasonable rankings, as a way to help identify outliers. I would be happy to tweet a request to the weirdos who follow me, if that would be helpful.

**Comment by adpi2** [comment]
The scoring function works approximately this way:
```
(
  Best match in (
    repo name * 6,
    organization * 5,
    github description * 4,
    github topics * 4
  ) +
  match in readme * 0.5
) *
LOG(number of github stars)
```

The weight of the readme is intentionally low because there is too much noisy information in it. Yet a readme that mentions the word `database` a dozen times, indicates that the library has something to do with databases.
